File: cracking_coding_interview/heap.py
class BinHeap:
    def __init__(self):
        self.heap_arr = [0]
        self.current_size = 0

    def bubble_up(self, i):
        while i // 2 > 0:
            if self.heap_arr[i] < self.heap_arr[i // 2]:  # parent is larger than child
                tmp = self.heap_arr[i // 2]
                self.heap_arr[i // 2] = self.heap_arr[i]
                self.heap_arr[i] = tmp
                i = i // 2

    def insert(self, k):
        self.heap_arr.append(k)
        self.current_size = self.current_size + 1
        self.bubble_up(self.current_size)

    def bubble_down(self, i):
        while (i * 2) <= self.current_size:
            mc = self.min_child(i)
            if self.heap_arr[i] > self.heap_arr[mc]:
                tmp = self.heap_arr[i]
                self.heap_arr[i] = self.heap_arr[mc]
                self.heap_arr[mc] = tmp
            i = mc

    def min_child(self, i):
        if i * 2 + 1 > self.current_size:
            return i * 2
        else:
            if self.heap_arr[i * 2] < self.heap_arr[i * 2 + 1]:
                return i * 2
            else:
                return i * 2 + 1

    def delete_min(self):
        item = self.heap_arr[1]
        self.heap_arr[1] = self.heap_arr[self.current_size]
        self.current_size = self.current_size - 1
        self.heap_arr.pop()
        self.bubble_down(1)
        return item

    def build_heap(self, arr):
        i = len(arr) // 2
        self.current_size = len(arr)
        self.heap_arr = [0] + arr[:]
        while (i > 0):
            self.bubble_down(i)
            i = i - 1

File: SkienaAlgorithmDesignManualJava/src/Heap.java
import java.util.ArrayList;
import java.util.Objects;

public class Heap<T extends Comparable<T>> {

    private ArrayList<T> arr;

    public Heap() {
        initializeArray();
    }

    public Heap(T[] list) {
        initializeArray();
        for (T elem : list) {
            insert(elem);
        }
    }

    private void initializeArray() {
        arr = new ArrayList<T>();
        arr.add(null);
    }

    public int size() {
        return arr.size() - 1;
    }

    public int parentIndex(int idx) {
        if (idx <= 1) {
            return -1;
        }
        return idx / 2; // implicitly does floor
    }

    public int youngChildIndex(int idx) {
        return 2 * idx;
    }

    public void insert(T elem) {
        arr.add(elem);
        bubbleUp(arr.size() - 1);
    }

    public T extractMin() {
        if (size() <= 0) {
            throw new java.lang.IllegalStateException("min value requested for empty priority queue!");
        }

        T min = arr.get(1);
        swap(1, size());
        arr.remove(size());
        bubbleDown(1);

        return min;
    }

    private void bubbleDown(int idx) {
        int leftChild = youngChildIndex(idx);
        int rightChild = leftChild + 1;
        int minIndex = idx;

        // Compare parent and left child.
        if (leftChild <= size() && arr.get(minIndex).compareTo(arr.get(leftChild)) > 0) {
            minIndex = leftChild;
        }
        // Compare the minimum with right child.
        if (rightChild <= size() && arr.get(minIndex).compareTo(arr.get(rightChild)) > 0) {
            minIndex = rightChild;
        }

        if (!Objects.equals(minIndex, idx)) {
            swap(minIndex, idx);
            bubbleDown(minIndex);
        }
    }

    private void bubbleUp(int idx) {
        if (parentIndex(idx) == -1) {
            return;
        }
        // parent should be larger than child in a heap
        if (arr.get(parentIndex(idx)).compareTo(arr.get(idx)) > 0) {
            swap(idx, parentIndex(idx));
            bubbleUp(parentIndex(idx));
        }
    }

    private void swap(int idx1, int idx2) {
        T temp = arr.get(idx1);
        arr.set(idx1, arr.get(idx2));
        arr.set(idx2, temp);
    }
}
